File: ShortestWordDistance2.java
class WordDistance {


    //Time Complexity: 0(nk) where n is the no of words in the array and k is the average length of the words
    //Space Complexity: 0(n)

    HashMap<String, List<Integer>> map; //a map to store the words in words dict and a list of indexes of it's occurences

    public WordDistance(String[] wordsDict) {

        map = new HashMap<>();

        for(int i = 0; i < wordsDict.length; i++){  //going over dict and storing each word to hashmap and its occurence
            String s = wordsDict[i];
            if(!map.containsKey(s)){
                map.put(s, new ArrayList<>());
            }
            map.get(s).add(i);
        }
    }

    public int shortest(String word1, String word2) {

        int dist = Integer.MAX_VALUE;   //calculates the minimum distance
        List<Integer> l1 = map.get(word1);  //to get the list of occurence of word 1
        List<Integer> l2 = map.get(word2);  //gets the list of occurence of word 2
        int p1 = 0; //pointer to navigate the l1 and l2 list
        int p2 = 0;

        while(p1 < l1.size() && p2 < l2.size()){    //till any one pointer goes out of bounds
            int idx1 = l1.get(p1);  //getting the 1st index of occurence of word1 and word 2 in list l1 and l2 respectively
            int idx2 = l2.get(p2);
            dist = Math.min(dist, Math.abs(idx1 - idx2));   //computing the distance between them
            if(idx1 < idx2){    //if 1st index is smaller, then I move the pointer in l1 to narrow the difference
                p1++;
            }
            else{   //else I move the pointer in l2 to narrow the difference
                p2++;
            }
        }

        return dist;    //return distance
    }
}

/**
 * Your WordDistance object will be instantiated and called as such:
 * WordDistance obj = new WordDistance(wordsDict);
 * int param_1 = obj.shortest(word1,word2);
 */

File: ShortestWordDistance2.py
class WordDistance:
    
    '''
    Time Complexity - O(N)
    Space Complexity - O(1)
    '''

    def __init__(self, words: List[str]):
        
        self.hmap={} # For storing all the index of occurences of a particular word
        for i,word in enumerate(words):
            
            if word in self.hmap:
                self.hmap[word].append(i)
            else:
                self.hmap[word]=[i] # key will be the word and value will be the list of indexes where the word occurs in the list
            
        

    def shortest(self, word1: str, word2: str) -> int:
        if len(word1)==0 or len(word2)==0:
            return 0
        
        minimum_distance=float("inf")
        
        list1=self.hmap[word1] # list of indexes where word 1 has occured
        list2=self.hmap[word2]# list of indexes of word2
        
        ptr1=0 #pointer for list1
        ptr2=0#pointer for list2
        # Example list1=[1,4,5] list2=[6,7] 
        
        while(ptr1<len(list1) and ptr2<len(list2)):
            index1=list1[ptr1]
            index2=list2[ptr2]
            
            minimum_distance = min(minimum_distance,abs(index1-index2))
            if index1>index2: # When value of list1 at ptr1 index is greater than index2--> Increment ptr2
                ptr2+=1
            else:
                ptr1+=1
                
        return minimum_distance
            
        


# Your WordDistance object will be instantiated and called as such:
# obj = WordDistance(words)
# param_1 = obj.shortest(word1,word2)